_posts/2018-12-05-exploit-model.md

```diff
@@ -0,0 +1,184 @@
+
+### Another model: LibOS with Nabla
+
+Let's compare the Nabla containers model with that of gvisor. The components are very similar, gvisor is a userspace unikernel, and Nabla has a Library OS (LibOS). This LibOS does a similar function - implementing the functionality of the kernel in userspace and only calling the kernel when necessary. 
+
+Right now, we don't distinguish between the application and the LibOS, and we treat the LibOS as untrusted. For the application and LibOS, we have a seccomp policy that only allows 7 syscalls, with only two file descriptions (a block device and tap device).
+
+```
++--------+              +---------+
+|  p1a   |              |   p2a   |                      APP
++-------------------+   +-------------------+
+|    nabla LibOS    |   |   nabla LibOS     |            LibOS
+|    (code exec)    |   |                   |            
++-------------------+   +-------------------+
+<<<   7 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+For nabla, the attack surface to obtain code exec in the kernel would need to be from these 7 syscalls.
+
+To summarize, here's a side-by-side comparison of the models:
+
+```
+      GVISOR                   NABLA
+      ------                   -----
+
++--------+              
+|  p1a   |                                              APP
++--------+              +---------+
+<<< 300 syscalls  >>>   |   p2a   |          
++-------------------+   +-------------------+
+|      gvisor       |   |   nabla LibOS     |           LibOS
+|   (code exec)     |   |                   |            
++-------------------+   +-------------------+
+<<<  83 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+
+
+
+## But it's a memory safe langauge!
+
+So gvisor does two things, it reduces the attack surface to the kernel, and has an protection boundary which guards against the inputs that can be invoked with the 83 syscalls (the guarding is implicit in a way that the logic of the implementation would constrain the total number of syscalls invokable by a well behaved gvisor). Therefore, more work needs to be done by the attacker to bypass these guards.
+
+Some argue that gvisor is written in golang, and golang is memory safe. I believe that golang is a safer language from the PL standpoint, but it is hard to guarantee that the implementation is true to the language semantics. Often, bugs in implementation or performance trade-offs (this is rampant in the hardware work with [Spectre](https://meltdownattack.com/) variant bugs) are the result of this.
```

Comment:
expand PL

Reply:
The relationship between this subsection to givsor's attack surface does not come it clearly. Maybe need to explicitly mention the message you want to give away here, beforehand.

Reply:
I think my main idea here is that it doesn't mean that because something is written in golang, that it is immune to exploits.

Reply:
I'm considering removing this section altogether, it is rather disjoint to the rest of the blogpost. All it's saying is that memory corruption is also a possibility in golang and should not be brushed off as impossible.